### PR TITLE
Convert promote double

### LINKED ISSUE #550: Unit conversion promote precision to double
The current implementation of unit conversion does something like this:
https://github.com/scipp/scipp/blob/20a43923fce75349c9e148b8264f4e7834ecfad1/neutron/convert.cpp#L36-L37

The problem here is that while we need to avoid `*=` to be able to broadcast, it also implies that the `dtype` can change. If the coord is `float`, the result is typically `double` since the beamline geometry information used to create `factor` is `double`.

I am not sure what the best way to avoid this is. Probably we need to explicitly convert `factor` to the same `dtype` as the coord.

Another option could be to avoid `*` but use a custom `transform` instead: A functor could force the second argument to be the same as the first, doing the type conversion during the transform, without the need for allocating convert temporaries.

### PR DESCRIPTION
**Pull request by igudich**
`astype` and `reciprocal` are introduced.
fixes #550

### CODE REVIEW

File: core/include/scipp/core/value_and_variance.h [hunk 1]
```diff
@@ -56,6 +56,12 @@ constexpr auto operator-(const ValueAndVariance<T> a) noexcept {
   return ValueAndVariance{-a.value, a.variance};
 }
 
+template <class T> constexpr auto reciprocal(const ValueAndVariance<T> &a) {
+  auto f = static_cast<T>(1) / a.value;
+  auto tmp = -f * a.variance / a.value;
+  return ValueAndVariance<T>(f, tmp * tmp);
```
Comment (SimonHeybrock) [inline core/include/scipp/core/value_and_variance.h]: `return static_cast<T>(1) / a;`?

File: core/include/scipp/core/variable.h [hunk 2]
```diff
@@ -889,6 +889,9 @@ operator/(T v, const units::Unit &unit) {
       {}, units::Unit(units::dimensionless) / unit, {v});
 }
 
+SCIPP_CORE_EXPORT Variable astype(const VariableConstProxy &var, DType type);
```
Comment (SimonHeybrock) [inline core/include/scipp/core/variable.h]:
```suggestion
SCIPP_CORE_EXPORT Variable astype(const VariableConstProxy &var, const DType type);
```

File: core/variable_binary_arithmetic.cpp [hunk 3]
```diff
@@ -279,4 +304,30 @@ Variable operator/(const double a, const VariableConstProxy &b_proxy) {
...
+                                     static_cast<T>(x.variance)};
+                               else
+                                 return static_cast<T>(x);
+                             }});
+    }
+  };
+  static Variable make(const VariableConstProxy &var, DType type) {
+    return CallDType<double, float, int64_t, int32_t, bool>::apply<Maker>(type,
+                                                                          var);
+  }
```
Comment (SimonHeybrock) [inline core/variable_binary_arithmetic.cpp]: Why is this wrapping function required? And why do we need the `struct MakeVariableWithType`?

File: neutron/convert.cpp [hunk 4]
```diff
@@ -12,6 +12,8 @@
 #include "scipp/neutron/beamline.h"
 #include "scipp/neutron/convert.h"
 
+#include <iostream>
+
```
Comment (SimonHeybrock) [inline neutron/convert.cpp]: Debugging leftover?

File: neutron/convert.cpp [hunk 5]
```diff
@@ -97,24 +101,26 @@ auto tofToEnergyConversionFactor(const Dataset &d) {
 
 Dataset tofToEnergy(Dataset &&d) {
   // 1. Compute conversion factor
-  const auto conversionFactor = tofToEnergyConversionFactor(d);
+  const auto conversionFactor =
+      astype(tofToEnergyConversionFactor(d), d.coords()[Dim::Tof].dtype());
```
Comment (SimonHeybrock) [inline neutron/convert.cpp]: I think you need to keep the conversion factor with its actual `dtype`. For event data there may be multiple things that are converted, and some may be `double` while others are `float`. I think the correct way to do this is as you did above in `convert_with_factor`.

File: neutron/convert.cpp [hunk 6]
```diff
@@ -128,7 +134,8 @@ Dataset tofToEnergy(Dataset &&d) {
 
 Dataset energyToTof(Dataset &&d) {
   // 1. Compute conversion factor
-  const auto conversionFactor = tofToEnergyConversionFactor(d);
+  const auto conversionFactor =
+      astype(tofToEnergyConversionFactor(d), d.coords()[Dim::Energy].dtype());
```
Comment (SimonHeybrock) [inline neutron/convert.cpp]: Same comment as above: I think we cannot convert the dtype of `conversionFactor`, but have to do it individually for each coord (global dense coord, and sparse coords).

File: neutron/test/convert_test.cpp [hunk 7]
```diff
@@ -469,6 +469,20 @@ TEST(Convert, Energy_to_Tof_Elastic) {
                      tof_original["density"].values<double>(), 1e-12));
 }
 
+TEST(Convert, convert_with_factor_type_promotion) {
+  Dataset tof = makeTofDataForUnitConversion();
+  tof.setCoord(Dim::Tof, makeVariable<float>({Dim::Tof, 4}, units::us,
+                                             {4000, 5000, 6100, 7300}));
+  auto dimsToConvert = std::vector{Dim::Energy};
+  for (auto &&d : dimsToConvert) {
```
Comment (SimonHeybrock) [inline neutron/test/convert_test.cpp]: Why this loop?

File: core/variable_binary_arithmetic.cpp (hunk 3 above)
Comment (igudich) [reply core/variable_binary_arithmetic.cpp]: To keep the corresponding functionality grouped in one instance I would add the types list for both `transform` and `CallDType` as
```c++
using types = std::tuple<....>;
```
they can't work with tuples.

File: core/variable_binary_arithmetic.cpp [hunk 8]
```diff
@@ -53,6 +55,29 @@ template <class T1, class T2> Variable plus(const T1 &a, const T2 &b) {
...
+public:
+  static constexpr bool value =
+      !std::is_same_v<void, decltype(detect(std::declval<T>()))>;
+};
+
+static constexpr auto reciprocal_ = [](const auto a_) {
+  if constexpr (hasReciprocal<std::decay_t<decltype(a_)>>::value)
+    return reciprocal(a_);
+  else
+    return static_cast<decltype(a_)>(1) / a_;
```
Comment (SimonHeybrock) [inline core/variable_binary_arithmetic.cpp]: Why is this fallback required? You are only calling `transform` with `float` and `double`?

File: neutron/test/convert_test.cpp [hunk 9]
```diff
@@ -469,6 +469,20 @@ TEST(Convert, Energy_to_Tof_Elastic) {
                      tof_original["density"].values<double>(), 1e-12));
 }
 
+TEST(Convert, convert_with_factor_type_promotion) {
+  Dataset tof = makeTofDataForUnitConversion();
+  tof.setCoord(Dim::Tof, makeVariable<float>({Dim::Tof, 4}, units::us,
+                                             {4000, 5000, 6100, 7300}));
+  auto dimsToConvert = std::vector{Dim::DSpacing, Dim::Wavelength, Dim::Energy};
```
Comment (SimonHeybrock) [inline neutron/test/convert_test.cpp]: Note that you could do it without introducing a named variable:

```cpp
for (auto &&d : {Dim::DSpacing, Dim::Wavelength, Dim::Energy}) {
```

File: core/variable_binary_arithmetic.cpp [hunk 10]
```diff
@@ -64,18 +64,18 @@ template <class T> class hasReciprocal {
       !std::is_same_v<void, decltype(detect(std::declval<T>()))>;
 };
```
Comment (SimonHeybrock) [inline core/variable_binary_arithmetic.cpp]: You can also remove this, and the addition of `reciprocal` for `ValueAndVariance`.

**Pull request merged**